File: tests/unit/xray/fixtures/csharp/advanced.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Example.Advanced
{
    // Records with positional and nominal syntax
    public record Point(double X, double Y)
    {
        public double DistanceTo(Point other) =>
            Math.Sqrt(Math.Pow(X - other.X, 2) + Math.Pow(Y - other.Y, 2));

        public static Point Origin => new(0, 0);
    }

    public record Circle(Point Center, double Radius) : IShape
    {
        public double Area => Math.PI * Radius * Radius;
        public double Perimeter => 2 * Math.PI * Radius;
    }

    public record Rectangle(Point TopLeft, double Width, double Height) : IShape
    {
        public double Area => Width * Height;
        public double Perimeter => 2 * (Width + Height);
        public bool IsSquare => Math.Abs(Width - Height) < double.Epsilon;
    }

    public interface IShape
    {
        double Area { get; }
        double Perimeter { get; }
    }

    // Pattern matching switch expression
    public static class ShapeAnalyzer
    {
        public static string Describe(IShape shape) => shape switch
        {
            Circle c when c.Radius > 100 => $"Large circle (r={c.Radius:F1})",
            Circle c => $"Circle (r={c.Radius:F1})",
            Rectangle { IsSquare: true } r => $"Square (side={r.Width:F1})",
            Rectangle r => $"Rectangle ({r.Width:F1}x{r.Height:F1})",
            null => throw new ArgumentNullException(nameof(shape)),
            _ => $"Unknown shape: {shape.GetType().Name}",
        };

        public static double TotalArea(IEnumerable<IShape> shapes) =>
            shapes.Sum(s => s.Area);

        public static IShape Largest(IEnumerable<IShape> shapes) =>
            shapes.MaxBy(s => s.Area)
            ?? throw new InvalidOperationException("No shapes provided");
    }

    // Generic result type
    public class Result<T>
    {
        private readonly T? _value;
        private readonly Exception? _error;

        private Result(T value) => _value = value;
        private Result(Exception error) => _error = error;

        public static Result<T> Ok(T value) => new(value);
        public static Result<T> Fail(Exception error) => new(error);

        public bool IsOk => _error is null;
        public T Value => IsOk ? _value! : throw new InvalidOperationException("Result is failure", _error);
        public Exception Error => !IsOk ? _error! : throw new InvalidOperationException("Result is success");

        public Result<TOut> Map<TOut>(Func<T, TOut> f) =>
            IsOk ? Result<TOut>.Ok(f(_value!)) : Result<TOut>.Fail(_error!);

        public T GetOrDefault(T defaultValue) => IsOk ? _value! : defaultValue;
    }

    // Fluent LINQ pipeline
    public static class DataPipeline
    {
        public static IEnumerable<(string Key, double Average, int Count)> Summarize(
            IEnumerable<(string Category, double Value)> data) =>
            data
                .GroupBy(x => x.Category)
                .Select(g => (
                    Key: g.Key,
                    Average: g.Average(x => x.Value),
                    Count: g.Count()
                ))
                .Where(s => s.Count > 1)
                .OrderByDescending(s => s.Average);
    }

    // Extension methods
    public static class Extensions
    {
        public static IEnumerable<T> WhereNotNull<T>(this IEnumerable<T?> source) where T : class =>
            source.Where(x => x is not null).Select(x => x!);

        public static TResult Pipe<T, TResult>(this T value, Func<T, TResult> fn) => fn(value);

        public static IEnumerable<T> Tap<T>(this IEnumerable<T> source, Action<T> action)
        {
            foreach (var item in source)
            {
                action(item);
                yield return item;
            }
        }
    }
}

File: tests/unit/xray/fixtures/java/advanced.java
package com.example.advanced;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Advanced Java features: sealed classes, records, pattern matching switch.
 */
public class Advanced {

    // Sealed interface hierarchy
    public sealed interface Shape permits Circle, Rectangle, Triangle {}

    public record Circle(double radius) implements Shape {
        public double area() { return Math.PI * radius * radius; }
    }

    public record Rectangle(double width, double height) implements Shape {
        public double area() { return width * height; }
    }

    public record Triangle(double base, double height) implements Shape {
        public double area() { return 0.5 * base * height; }
    }

    // Pattern matching switch (Java 17+)
    public static String describeShape(Shape shape) {
        return switch (shape) {
            case Circle c -> "Circle with radius " + c.radius();
            case Rectangle r when r.width() == r.height() -> "Square with side " + r.width();
            case Rectangle r -> "Rectangle " + r.width() + "x" + r.height();
            case Triangle t -> "Triangle base=" + t.base() + " height=" + t.height();
        };
    }

    public static double computeArea(Shape shape) {
        return switch (shape) {
            case Circle c -> c.area();
            case Rectangle r -> r.area();
            case Triangle t -> t.area();
        };
    }

    // Generic bounded type
    public static <T extends Shape & Comparable<T>> T findLargest(List<T> shapes) {
        return shapes.stream()
                .reduce((a, b) -> a.compareTo(b) >= 0 ? a : b)
                .orElseThrow(() -> new IllegalArgumentException("Empty list"));
    }

    // Text blocks (Java 15+)
    public static final String JSON_TEMPLATE = """
            {
                "type": "%s",
                "area": %.2f,
                "perimeter": %.2f
            }
            """;

    // Nested records
    public record Point(double x, double y) {
        public double distanceTo(Point other) {
            double dx = this.x - other.x;
            double dy = this.y - other.y;
            return Math.sqrt(dx * dx + dy * dy);
        }
    }

    public record Line(Point start, Point end) {
        public double length() { return start.distanceTo(end); }
        public Point midpoint() {
            return new Point((start.x() + end.x()) / 2, (start.y() + end.y()) / 2);
        }
    }

    // Lambda with method references
    public static List<String> formatShapes(List<Shape> shapes) {
        return shapes.stream()
                .map(Advanced::describeShape)
                .sorted(String::compareTo)
                .collect(Collectors.toList());
    }

    // Optional chaining
    public static Optional<Double> safeArea(Shape shape) {
        return Optional.ofNullable(shape)
                .map(Advanced::computeArea)
                .filter(area -> area > 0);
    }

    // Nested lambda capturing outer scope
    public static Function<Double, Shape> circleFactory() {
        double defaultRadius = 1.0;
        return radius -> {
            double effective = radius > 0 ? radius : defaultRadius;
            return new Circle(effective);
        };
    }

    public static void main(String[] args) {
        List<Shape> shapes = List.of(
                new Circle(5.0),
                new Rectangle(3.0, 4.0),
                new Triangle(6.0, 8.0),
                new Rectangle(2.0, 2.0)
        );

        shapes.stream()
                .map(s -> String.format(JSON_TEMPLATE, s.getClass().getSimpleName(),
                        computeArea(s), 0.0))
                .forEach(System.out::println);

        formatShapes(shapes).forEach(System.out::println);
    }
}
